api/apiclient_test.go

```diff
@@ -1205,6 +1212,177 @@ func (s *apiclientSuite) TestLoginIncompatibleClient(c *gc.C) {
+		err := fmt.Errorf("bad read")
+		return api.NewFakeStreamReader(&badReader{err}), nil
+	})
+	reader, err := s.APIState.ConnectStream("/", nil)
+	c.Assert(err, gc.ErrorMatches, "unable to read initial response: bad read")
+	c.Assert(reader, gc.IsNil)
+}
+
+// badReader raises err when Read is called.
+type badReader struct {
+	err error
+}
+
+func (r *badReader) Read(p []byte) (n int, err error) {
+	return 0, r.err
+}
+
+func (s *apiclientSuite) TestConnectControllerStreamRejectsRelativePaths(c *gc.C) {
+	reader, err := s.APIState.ConnectControllerStream("foo", nil, nil)
+	c.Assert(err, gc.ErrorMatches, `path "foo" is not absolute`)
+	c.Assert(reader, gc.IsNil)
+}
+
+func (s *apiclientSuite) TestConnectControllerStreamRejectsModelPaths(c *gc.C) {
+	reader, err := s.APIState.ConnectControllerStream("/model/foo", nil, nil)
+	c.Assert(err, gc.ErrorMatches, `path "/model/foo" is model-specific`)
+	c.Assert(reader, gc.IsNil)
+}
+
+func (s *apiclientSuite) TestConnectControllerStreamAppliesHeaders(c *gc.C) {
+	catcher := api.UrlCatcher{}
+	headers := http.Header{}
+	headers.Add("thomas", "cromwell")
+	headers.Add("anne", "boleyn")
+	s.PatchValue(&api.WebsocketDial, catcher.RecordLocation)
+
+	_, err := s.APIState.ConnectControllerStream("/something", nil, headers)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(catcher.Headers().Get("thomas"), gc.Equals, "cromwell")
+	c.Assert(catcher.Headers().Get("anne"), gc.Equals, "boleyn")
+}
+
+func (s *apiclientSuite) TestWatchDebugLogParamsEncoded(c *gc.C) {
+	catcher := api.UrlCatcher{}
+	s.PatchValue(&api.WebsocketDial, catcher.RecordLocation)
+
+	params := common.DebugLogParams{
+		IncludeEntity: []string{"a", "b"},
+		IncludeModule: []string{"c", "d"},
+		IncludeLabel:  []string{"e", "f"},
+		ExcludeEntity: []string{"g", "h"},
+		ExcludeModule: []string{"i", "j"},
+		ExcludeLabel:  []string{"k", "l"},
+		Limit:         100,
+		Backlog:       200,
+		Level:         loggo.ERROR,
+		Replay:        true,
+		NoTail:        true,
+		StartTime:     time.Date(2016, 11, 30, 11, 48, 0, 100, time.UTC),
+	}
+
+	urlValues := url.Values{
+		"includeEntity": params.IncludeEntity,
+		"includeModule": params.IncludeModule,
+		"includeLabel":  params.IncludeLabel,
+		"excludeEntity": params.ExcludeEntity,
+		"excludeModule": params.ExcludeModule,
+		"excludeLabel":  params.ExcludeLabel,
+		"maxLines":      {"100"},
+		"backlog":       {"200"},
+		"level":         {"ERROR"},
+		"replay":        {"true"},
+		"noTail":        {"true"},
+		"startTime":     {"2016-11-30T11:48:00.0000001Z"},
+	}
+
+	client := apiclient.NewClient(s.APIState)
+	_, err := client.WatchDebugLog(params)
+	c.Assert(err, jc.ErrorIsNil)
+
+	connectURL, err := url.Parse(catcher.Location())
+	c.Assert(err, jc.ErrorIsNil)
+
+	values := connectURL.Query()
+	c.Assert(values, jc.DeepEquals, urlValues)
+}
+
+func (s *apiclientSuite) TestWatchDebugLogConnected(c *gc.C) {
```

Comment:
Why remove the use of gomock from this test when it moved?

Reply:
I didn't realize we were gonna gomock things here as well, I'll just keep it like this for now, we'll gomock it out with the others later on.